include/NeoFOAM/auxiliary/readers.hpp

```diff
@@ -176,48 +159,65 @@ auto readSurfaceBoundaryConditions(
     return bcs;
 }
 
-template<typename FoamType>
-auto constructSurfaceField(
-    const NeoN::Executor exec,
-    const NeoN::UnstructuredMesh& nfMesh,
-    const FoamType& in
-)
+template <class FoamFieldType>
+auto constructFrom(const NeoN::Executor exec,
+                   const NeoN::UnstructuredMesh& nfMesh,
+                   const FoamFieldType& in)
 {
-    using type_container_t = typename TypeMap<FoamType>::container_type;
-    using type_primitive_t = typename TypeMap<FoamType>::mapped_type;
-    using foam_primitive_t = typename FoamType::cmptType;
+    using ContainerType = typename TypeMap<FoamFieldType>::container_type;
+    using MappedType    = typename TypeMap<FoamFieldType>::mapped_type;
 
-    type_container_t
-        out(exec, in.name(), nfMesh, std::move(readSurfaceBoundaryConditions(nfMesh, in)));
+    if constexpr (NeoFOAM::detail::kIsVolumeFieldV<ContainerType>) {
+        ContainerType out(exec, in.name(), nfMesh, readVolBoundaryConditions(nfMesh, in));
+        out.internalVector() = fromFoamField(exec, in.primitiveField());
+        out.correctBoundaryConditions();
+        return out;
+    } else if constexpr (NeoFOAM::detail::kIsSurfaceFieldV<ContainerType>) {
+        using FoamComponentType = typename FoamFieldType::cmptType;
 
-    Foam::Field<foam_primitive_t> flattenedField(out.internalVector().size());
-    size_t nInternal = nfMesh.nInternalFaces();
+        ContainerType out(exec, in.name(), nfMesh, readSurfaceBoundaryConditions(nfMesh, in));
 
-    forAll(in, facei)
-    {
-        flattenedField[facei] = convert(in[facei]);
-    }
+        const std::size_t nInt   = nfMesh.nInternalFaces();
+        const std::size_t nBnd   = nfMesh.boundaryMesh().offset().back();
+        const std::size_t nFaces = nInt+nBnd;
 
-    Foam::label idx = nInternal;
-    Foam::Field<foam_primitive_t> bvalue(out.internalVector().size());
-    forAll(in.boundaryField(), patchi)
-    {
-        const Foam::fvsPatchField<foam_primitive_t>& pin = in.boundaryField()[patchi];
+	NF_DINFO("Internal: "+std::to_string(nInt)+", Boundary: "+std::to_string(nBnd)+", nFaces: "+std::to_string(nFaces));
```

Comment:
Will this work out of the box?
```suggestion
	NeoN::Logging::info("Internal: {}, Boundary: {}, nFaces: {}", nInt, nBnd, nFaces);
```

Reply:
yes, but the we might want to have it as `NeoN::Logging::debug` which is not yet available. with info it will always print